File: src/main/java/com/common/lib/api/controller/CrudController.java
package com.common.lib.api.controller;

import com.common.lib.utils.PlantillaResponse;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

import java.util.Map;

/**
 * Interfaz CRUD reactiva generica.
 * @param <RES> DTO de respuesta
 * @param <RQ> DTO de request
 */
public interface CrudController<RES, RQ> {

    @PostMapping("/add")
    Mono<ResponseEntity<PlantillaResponse<RES>>> add(
        @RequestBody RQ request,
        @RequestHeader HttpHeaders headers
    );

    @GetMapping("/all")
    Mono<ResponseEntity<PlantillaResponse<RES>>> all(
        @RequestHeader HttpHeaders headers,
        @RequestParam(required = false) Map<String, String> filters,
        @RequestParam(required = false) Object id
    );

    @PutMapping("/update")
    Mono<ResponseEntity<PlantillaResponse<RES>>> update(
        @RequestBody RQ request,
        @RequestHeader HttpHeaders headers,
        @RequestParam Object id
    );

    @DeleteMapping("/delete")
    Mono<ResponseEntity<PlantillaResponse<RES>>> delete(
        @RequestHeader HttpHeaders headers,
        @RequestParam(required = false) Object id
    );
}

File: src/main/java/com/common/lib/api/controller/CrudController.kt
package com.common.lib.api.controller

import com.common.lib.utils.PlantillaResponse
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Mono

/**
 * Interfaz para los controladores CRUD, equivalente a la interfaz TypeScript.
 * @param RES Clase de respuesta
 * @param RQ Clase de solicitud
 */
interface CrudController<RES, RQ> {

    @PostMapping("/add")
    fun add(
        @RequestBody request: RQ,
        @RequestHeader("id") id: String?,
        @RequestHeader("ip") ip: String,
        @RequestHeader("dominio") dominio: String,
        @RequestHeader("usuario") usuario: String,
        @RequestHeader("idbusiness", required = false) idbusiness: Int?,
        @RequestHeader("proceso") proceso: String,
        @RequestHeader("topic") topic: String,
        @RequestHeader("token", required = false) token: String?
    ): ResponseEntity<PlantillaResponse<RES>>

    @GetMapping("/all")
    fun all(
        @RequestHeader("id") id: String?,
        @RequestHeader("ip") ip: String,
        @RequestHeader("dominio") dominio: String,
        @RequestHeader("usuario") usuario: String,
        @RequestHeader("idbusiness", required = false) idbusiness: Int?,
        @RequestHeader("proceso") proceso: String,
        @RequestHeader("topic") topic: String,
        @RequestHeader("token", required = false) token: String?,
        @RequestParam(required = false) filters: Map<String, String>?
    ): ResponseEntity<PlantillaResponse<RES>>

    @PutMapping("/update")
    fun update(
        @RequestBody request: RQ,
        @RequestHeader("id") id: String?,
        @RequestHeader("ip") ip: String,
        @RequestHeader("dominio") dominio: String,
        @RequestHeader("usuario") usuario: String,
        @RequestHeader("idbusiness", required = false) idbusiness: Int?,
        @RequestHeader("proceso") proceso: String,
        @RequestHeader("topic") topic: String,
        @RequestHeader("token", required = false) token: String?
    ): Mono<PlantillaResponse<RES>>

    @DeleteMapping("/delete/{id}")
    fun delete(
        @PathVariable id: String,
        @RequestHeader("ip") ip: String,
        @RequestHeader("dominio") dominio: String,
        @RequestHeader("usuario") usuario: String,
        @RequestHeader("idbusiness", required = false) idbusiness: Int?,
        @RequestHeader("proceso") proceso: String,
        @RequestHeader("topic") topic: String,
        @RequestHeader("token", required = false) token: String?
    ): ResponseEntity<PlantillaResponse<RES>>
}
